cocotb/drivers/amba.py

```diff
@@ -129,86 +251,229 @@ def write(
         """
-        if sync:
-            yield RisingEdge(self.clock)
 
-        c_addr = cocotb.fork(self._send_write_address(address,
-                                                      delay=address_latency))
-        c_data = cocotb.fork(self._send_write_data(value,
-                                                   byte_enable=byte_enable,
-                                                   delay=data_latency))
+        try:
+            iter(value)
+        except TypeError:
+            value = (value,)    # If value is not an iterable, make it
 
-        if c_addr:
-            yield c_addr.join()
-        if c_data:
-            yield c_data.join()
+        if size is None:
+            size = len(self.bus.WDATA) // 8
+        else:
+            AXI4Master._check_size(size, len(self.bus.WDATA) // 8)
 
-        # Wait for the response
-        while True:
-            yield ReadOnly()
-            if self.bus.BVALID.value and self.bus.BREADY.value:
-                result = self.bus.BRESP.value
-                break
-            yield RisingEdge(self.clock)
+        AXI4Master._check_length(len(value), burst)
+        AXI4Master._check_4kB_boundary_crossing(address, burst, size,
+                                                len(value))
 
-        yield RisingEdge(self.clock)
+        write_address = self._send_write_address(address, len(value), burst,
+                                                 size, address_latency, sync)
 
-        if int(result):
-            raise AXIProtocolError("Write to address 0x%08x failed with BRESP: %d"
-                                   % (address, int(result)))
+        write_data = self._send_write_data(value, data_latency, byte_enable,
+                                           sync)
 
-        return result
+        await Combine(write_address, write_data)
+
+        async with self.write_response_busy:
+            # Wait for the response
+            while True:
+                await ReadOnly()
+                if self.bus.BVALID.value and self.bus.BREADY.value:
+                    result = AXIxRESP(self.bus.BRESP.value.integer)
+                    break
+                await RisingEdge(self.clock)
+
+            await RisingEdge(self.clock)
+
+            if result is not AXIxRESP.OKAY:
+                err_msg = "Write to address {0:#x}"
+                if len(value) != 1:
+                    err_msg += " ({1} beats, {2} burst)"
+                err_msg += " failed with BRESP: {3} ({4})"
+
+                raise AXIProtocolError(
+                    err_msg.format(address, len(value), burst.name,
+                                   result.value, result.name), result)
+
+            return result
 
     @cocotb.coroutine
-    def read(self, address: int, sync: bool = True) -> BinaryValue:
+    async def read(self, address, length=1, size=None, burst=AXIBurst.INCR,
+                   return_rresp=False, sync=True):
```

Comment:
I'd recommend making these keyword only. I think `length` is just about obvious enough that `bus.read(0xFF0000, 4)` is worth allowing as a shorthand for `bus.read(0xFF0000, length=4)`, but for all the other arguments explicit names seem wise.
```suggestion
    async def read(self, address, length=1, *, size=None, burst=AXIBurst.INCR,
                   return_rresp=False, sync=True):
```

Reply:
Or actually, while you're here, could add type annotations:
```suggestion
    async def read(self, address: int, length: int = 1, *
                   size: Optional[int] = None,
                   burst: AXIBurst = AXIBurst.INCR,
                   return_rresp: bool = False,
                   sync: bool=True):
```